src/univariate/discrete/zeroinflatedpoisson.jl

```diff
@@ -0,0 +1,178 @@
+```
+Zeros may arise from this process, an additional Bernoulli process, where the probability of observing an excess zero is given as p:
+```math
+P(X = 0) = p + (1 - p) e^{-\\lambda}
+```
+As p approaches 0, the distribution tends toward Poisson(λ).
+```julia
+ZeroInflatedPoisson()        # Zero-Inflated Poisson distribution with rate parameter 1, and probability of observing a zero 0.5
+ZeroInflatedPoisson(λ)       # ZeroInflatedPoisson distribution with rate parameter λ, and probability of observing a zero 0.5
+params(d)                    # Get the parameters, i.e. (λ, p)
+mean(d)                      # Get the mean of the mixture distribution
+var(d)                       # Get the variance of the mixture distribution
+```
+External links:
+* [Zero-inflated Poisson Regression on UCLA IDRE Statistical Consulting](https://stats.idre.ucla.edu/stata/dae/zero-inflated-poisson-regression/)
+* [Zero-inflated model on Wikipedia](https://en.wikipedia.org/wiki/Zero-inflated_model)
+* McElreath, R. (2020). Statistical Rethinking: A Bayesian Course with Examples in R and Stan (2nd ed.). Chapman and Hall/CRC. https://doi.org/10.1201/9780429029608
+
+"""
+struct ZeroInflatedPoisson{T<:Real} <: DiscreteUnivariateDistribution
+  λ::T
+  p::T
+
+  function ZeroInflatedPoisson{T}(λ::T, p::T) where {T <: Real}
+      return new{T}(λ, p)
+    end
+end
+
+function ZeroInflatedPoisson(λ::T, p::T; check_args = true) where {T <: Real}
+  if check_args
+    @check_args(Poisson, λ >= zero(λ))
+    @check_args(ZeroInflatedPoisson, zero(p) <= p <= one(p))
+  end
+  return ZeroInflatedPoisson{T}(λ, p)
+end
+
+ZeroInflatedPoisson(λ::Real, p::Real) = ZeroInflatedPoisson(promote(λ, p)...)
+ZeroInflatedPoisson(λ::Integer, p::Integer) = ZeroInflatedPoisson(float(λ), float(p))
+ZeroInflatedPoisson(λ::Real) = ZeroInflatedPoisson(λ, 0.0)
+ZeroInflatedPoisson() = ZeroInflatedPoisson(1.0, 0.0, check_args = false)
+
+@distr_support ZeroInflatedPoisson 0 (d.λ == zero(typeof(d.λ)) ? 0 : Inf)
+
+### Statistics
+
+mean(d::ZeroInflatedPoisson) = (1 - d.p) * d.λ
+
+var(d::ZeroInflatedPoisson) = d.λ * (1 - d.p) * (1 + d.p * d.λ)
+
+#### Conversions
+
+function convert(::Type{ZeroInflatedPoisson{T}}, λ::Real, p::Real) where {T<:Real}
+  return ZeroInflatedPoisson(T(λ), T(p))
+end
+
+function convert(::Type{ZeroInflatedPoisson{T}}, d::ZeroInflatedPoisson{S}) where {T <: Real, S <: Real}
+  return ZeroInflatedPoisson(T(d.λ), T(d.p), check_args = false)
+end
+
+#### Parameters
+
+import StatsBase:params
+
+params(d::ZeroInflatedPoisson) = (d.λ, d.p,)
+partype(::ZeroInflatedPoisson{T}) where {T} = T
+
+rate(d::ZeroInflatedPoisson) = d.λ
+excessprob(d::ZeroInflatedPoisson) = d.p
+
+#### Evaluation
+
+function logpdf(d::ZeroInflatedPoisson, y::Real)
+  LL = 0.0
+  if y == 0
+    LLs = zeros(typeof(log(d.λ)), 2)
+    LLs[1] = log(d.p)
+    LLs[2] = log1p(-d.p) - d.λ
+    LL = logsumexp(LLs)
+  else
+    LL = log1p(-d.p) + logpdf(Poisson(d.λ), y)
+  end
+  return LL
+end
```

Comment:
This function changes the type of `LL` which should be avoided and can be simplified:
```suggestion
function logpdf(d::ZeroInflatedPoisson, y::Real)
    lp = if iszero(y)
        logaddexp(log(d.p), log1p(-d.p) - d.λ)
    else
        log1p(-d.p) + logpdf(Poisson(d.λ), y)
    end
    return lp
end
```

Reply:
Great -- thank you.